go/services/adr/client.go

```diff
@@ -0,0 +1,539 @@
+	notifyOnAssetUpdateInvoker                         *adrbaseservice.NotifyOnAssetUpdateCommandInvoker
+	getAssetEndpointProfileCommandInvoker              *adrbaseservice.GetAssetEndpointProfileCommandInvoker
+	updateAssetEndpointProfileStatusCommandInvoker     *adrbaseservice.UpdateAssetEndpointProfileStatusCommandInvoker
+	createDiscoveredAssetEndpointProfileCommandInvoker *aeptypeservice.CreateDiscoveredAssetEndpointProfileCommandInvoker
+	createDetectedAssetCommandInvoker                  *adrbaseservice.CreateDetectedAssetCommandInvoker
+	getAssetCommandInvoker                             *adrbaseservice.GetAssetCommandInvoker
+	updateAssetStatusCommandInvoker                    *adrbaseservice.UpdateAssetStatusCommandInvoker
+}
+
+// ClientOption allows setting optional parameters on the Client.
+type ClientOption func(*Client)
+
+// WithLogger sets a logger for the client.
+func WithLogger(logger *slog.Logger) ClientOption {
+	return func(c *Client) {
+		c.logger = logger
+	}
+}
+
+// WithAssetUpdateHandler sets a handler for asset update events.
+func WithAssetUpdateHandler(
+	handler func(aepName string, asset *Asset) error,
+) ClientOption {
+	return func(c *Client) {
+		c.onAssetUpdate = handler
+	}
+}
+
+// WithAepUpdateHandler sets a handler for asset endpoint profile update events.
+func WithAepUpdateHandler(
+	handler func(aepName string, profile *AssetEndpointProfile) error,
+) ClientOption {
+	return func(c *Client) {
+		c.onAepUpdate = handler
+	}
+}
+
+// New creates a new ADR client.
+func New(
+	app *protocol.Application,
+	client protocol.MqttClient,
+	opts ...ClientOption,
+) (*Client, error) {
+	notifyInvoker, err := adrbaseservice.NewNotifyOnAssetEndpointProfileUpdateCommandInvoker(
+		app,
+		client,
+		"adr/v1/assetendpointprofiles/{aepName}/update",
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	c := &Client{
+		logger:         slog.Default(),
+		protocol:       app,
+		mqtt:           client,
+		observedAeps:   make(map[string]struct{}),
+		observedAssets: make(map[string]struct{}),
+		notifyOnAssetEndpointProfileUpdateInvoker: notifyInvoker,
+	}
+
+	for _, opt := range opts {
+		opt(c)
+	}
+
+	// Setup telemetry receivers
+	assetUpdateReceiver, err := protocol.NewTelemetryReceiver(
+		app,
+		client,
+		protocol.JSON[Asset]{},
+		"adr/v1/assets/{aepName}/update",
+		c.handleAssetUpdateTelemetry,
+	)
+	if err != nil {
+		return nil, err
+	}
+	c.listeners = append(c.listeners, assetUpdateReceiver)
+
+	aepUpdateReceiver, err := protocol.NewTelemetryReceiver(
+		app,
+		client,
+		protocol.JSON[AssetEndpointProfile]{},
+		"adr/v1/assetendpointprofiles/{aepName}/update",
+		c.handleAepUpdateTelemetry,
+	)
+	if err != nil {
+		return nil, err
+	}
+	c.listeners = append(c.listeners, aepUpdateReceiver)
+
+	return c, nil
+}
+
+// Close releases resources associated with the client.
+func (c *Client) Close() error {
+	var err error
+	c.closeOnce.Do(func() {
```

Comment:
nit: `c.listeners.Close()` should already be idempotent, so the `Once` shouldn't be necessary.